common/src/main/java/com/wynntils/features/user/overlays/map/MiniMapOverlayFeature.java

```diff
@@ -100,33 +110,131 @@ public void render(PoseStack poseStack, float partialTicks, Window window) {
-                RenderUtils.rotatePose(poseStack, centerX, centerZ, 180 - yRot);
+            if (WebManager.isMapLoaded()) {
+                renderMapQuad(map, poseStack, centerX, centerZ, textureX, textureZ, width, height);
             }
 
-            renderMapQuad(map, poseStack, centerX, centerZ, textureX, textureZ, width, height);
+            // TODO minimap icons
+
+            // TODO compass icon
+
+            // cursor
+            renderCursor(poseStack, centerX, centerZ);
+
+            // disable mask & render border
+            switch (maskType) {
+                case Rectangular -> {
+                    RenderSystem.disableScissor();
+                    renderRectangularMapBorder(poseStack, renderX, renderY, width, height);
+                }
+                    // case Circle -> {
+                    // TODO
+                    // renderCircularMapBorder();
+                    // }
+            }
+
+            // Directional Text
+            renderCardinalDirections(poseStack, width, height, centerX, centerZ);
+
+            // Coordinates
+            if (showCoords) {
+                String coords = String.format(
+                        "%s, %s, %s",
+                        (int) McUtils.player().getX(), (int) McUtils.player().getY(), (int)
+                                McUtils.player().getZ());
+
+                FontRenderer.getInstance()
+                        .renderText(
+                                poseStack,
+                                centerX,
+                                renderY + height + 10 * height / DEFAULT_SIZE,
+                                new TextRenderTask(
+                                        coords,
+                                        TextRenderSetting.CENTERED.withTextShadow(FontRenderer.TextShadow.OUTLINE)));
+            }
+        }
+
+        private void renderCardinalDirections(
+                PoseStack poseStack, float width, float height, float centerX, float centerZ) {
+            if (showCompass == CompassRenderType.None) return;
+
+            float northDX;
+            float northDY;
 
-            // disable rotation if necessary
             if (followPlayerRotation) {
-                poseStack.popPose();
+                float yawRadians = (float) Math.toRadians(McUtils.player().getYRot());
+                northDX = (float) StrictMath.sin(yawRadians);
+                northDY = (float) StrictMath.cos(yawRadians);
+                if (maskType == MapMaskType.Rectangular) {
+                    // Scale as necessary
+                    double toSquareScaleNorth = Math.min(width / Math.abs(northDX), height / Math.abs(northDY)) / 2;
+                    northDX *= toSquareScaleNorth;
+                    northDY *= toSquareScaleNorth;
+                }
+            } else {
+                northDX = 0;
+                northDY = -height / 2;
             }
 
-            // TODO minimap icons
+            FontRenderer.getInstance()
```

Comment:
For consistency, can you move the "N" rendering down to after the calculations, and group it with the other cardinal letter rendering?

Reply:
... and possibly fold the eastDX/DY calculation in with the northDX/DY calculation